File: src/network/netlink.rs
use std::io;

const Q_SIZE: usize = 4;

pub const IPADD: u8 = 20;
pub const IPRMV: u8 = 21;

#[derive(Clone, Debug, Default)]
#[repr(C)]
pub struct IfModMsg {
	pub devidx: u32,
	pub modop: u8, // enum (IPADD, IPRMV)
	pub ipv4: u32,
}

#[link(name="netlink")]
extern {
	fn nlsock_open() -> *mut u8;
	fn nlsock_close(ptr: *mut u8) -> i32;
	fn nlsock_recv(sock: *mut u8, vec: *mut IfModMsg, veclen: usize) -> i32;
}

pub struct NlSock {
	// Rust will not manipulate the C structure,
	// so abstract its implementation
	ptr: *mut u8,
	buf: [IfModMsg; Q_SIZE],
}

impl NlSock {
	pub fn recvmsg<'a>(&'a mut self) -> Vec<&'a IfModMsg> {
		let msgcount = unsafe {
			nlsock_recv(
				self.ptr,
				&mut self.buf as *mut IfModMsg,
				Q_SIZE
			)
		};

		if msgcount == -1 {
			panic!(
				"Failed to read netlink messages: {}",
				io::Error::last_os_error()
			);
		} else if msgcount < 0 {
			panic!("Failed to read netlink messages");
		}

		let mut vec: Vec<&'a IfModMsg> = Vec::new();
		for i in 0..msgcount {
			vec.push(&self.buf[i as usize]);
		}

		return vec;
	}

	pub fn new() -> NlSock {
		let ptr = unsafe { nlsock_open() };
		if ptr as usize == 0 {
			panic!("Failed to allocate netlink socket");
		}

		return NlSock {
			ptr,
			buf: Default::default(),
		}
	}
}

impl Drop for NlSock {
	fn drop(&mut self) {
		if unsafe { nlsock_close(self.ptr) } < 0 {
			panic!("Failed to deallocate socket");
		}
	}
}

File: src/network/netlink.c
#define _GNU_SOURCE

#include <errno.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <linux/netlink.h>
#include <linux/rtnetlink.h>
#include <sys/socket.h>
#include <sys/uio.h>

const unsigned IPADD = RTM_NEWADDR;
const unsigned IPRMV = RTM_DELADDR;
const size_t BUFSIZE = 256;

struct ifmodmsg {
	unsigned devidx;
	char modop;
	unsigned ipv4;
};

struct nlsock {
	int fd;
	struct msghdr bufh;
	struct sockaddr_nl addr;
	struct iovec bufv;
	char *buf;
};

/**
 * Opens a new nlsock
 *
 * @returns
 *     a newly allocated socket if successful
 *     null if an error occurs, in which case errno should be set
 */
struct nlsock* nlsock_open() {
	int fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);

	if (fd < 0) {
		return NULL;
	}

	struct nlsock* sock = (struct nlsock*)malloc(
		sizeof(struct nlsock)+BUFSIZE
	);
	memset(sock, 0, sizeof(struct nlsock)+BUFSIZE);

	sock->fd = fd;

	sock->addr.nl_family = AF_NETLINK;
	sock->addr.nl_groups = RTMGRP_IPV4_IFADDR;
	sock->addr.nl_pid = gettid() << 16 | getpid();

	sock->bufh.msg_name = &sock->addr;
	sock->bufh.msg_namelen = sizeof(struct sockaddr_nl);
	sock->bufh.msg_iov = &sock->bufv;
	sock->bufh.msg_iovlen = 1;

	sock->bufv.iov_base = (char*)sock + sizeof(struct nlsock);
	sock->bufv.iov_len = BUFSIZE;

	if (
		bind(
			sock->fd, (struct sockaddr*)&sock->addr, sizeof(struct sockaddr_nl)
		) < 0
	) {
		goto panic;
	}

	if (fcntl(sock->fd, F_SETFL, O_NONBLOCK) < 0) {
		goto panic;
	}

	return sock;

panic:
	close(sock->fd);
	free(sock);
	return NULL;
}

/**
 * Closes an open nlsock
 *
 * @returns
 *     0 if successful
 *     -1 if an error occurs, in which case errno should be set
 */
int nlsock_close(struct nlsock* sock) {
	if (close(sock->fd) < 0) {
		return -1;
	}

	free(sock);
	return 0;
}

/**
 * Reads buffered messages,
 * storing them in vec.
 *
 * @param bufv the buffer to read from
 * @param vec the vector on which to store messages
 * @param veclen the number of messages that vec can hold
 *
 * @returns
 *     the number of messages received, if the operation is successful
 *     -1 if a system error occurs, in which case errno should be set
 *     -2 if vec is not large enough to hold all received messages
 *     -3 if any message includes a non-IPv4 address
 *     -4 if any message includes more than one address for
 *     a single interface
 */
int readmsgs(
	struct iovec *bufv,
	size_t buflen,
	struct ifmodmsg *vec,
	size_t veclen
) {
	struct nlmsghdr *it = (struct nlmsghdr*)bufv;
	int msgs = 0;

	while (NLMSG_OK(it, buflen)) {
		if (it->nlmsg_type == NLMSG_ERROR) {
			return -1;
		} else if (msgs == veclen) {
			return -2;
		} else if (it->nlmsg_type == IPADD || it->nlmsg_type == IPRMV) {
			struct ifaddrmsg *addr = NLMSG_DATA(it);

			struct rtattr *rta = (struct rtattr *)(addr + 1);
			int rtalen = rta->rta_len;

			if (rta->rta_type != IFA_ADDRESS) {
				return -3;
			}

			vec[msgs].devidx = addr->ifa_index;
			vec[msgs].modop = it->nlmsg_type;
			vec[msgs].ipv4 = *((unsigned*)RTA_DATA(rta));

			RTA_NEXT(rta, rtalen);
			if (RTA_OK(rta, rtalen)) {
				return -4;
			}

			msgs++;
		}

		NLMSG_NEXT(it, buflen);
	}

	return msgs;
}

/**
 * Receives messages from the socket,
 * storing them in vec.
 *
 * @param sock the socket to read from
 * @param vec the vector on which to store the messages
 * @param veclen the number of messages that vec can hold
 *
 * @returns the same values as readmsgs()
 */
int nlsock_recv(
	struct nlsock *sock,
	struct ifmodmsg *vec,
	size_t veclen
) {
	ssize_t buflen = recvmsg(sock->fd, &sock->bufh, 0);

	if (buflen < 0) {
		if (errno == EWOULDBLOCK) {
			return 0;
		} else {
			return -1;
		}
	}

	return readmsgs(sock->bufv.iov_base, buflen, vec, veclen);
}
